### PR TITLE
CNI: add fallback logic if no ip address references sandboxed interface

### PR DESCRIPTION
**Pull request by nickethier**
This PR adds a branch of logic to how the alloc's IP address is discovered for CNI networks. Some CNI plugins don't reference an interface index in the result response. This leads to go-cni, the library we use for CNI, to associate the address with the default interface.

This logic will attempt to cycle through the interfaces and find the first address if there is no address found for the sandboxed interface.

Some additional debug log statements are also included to aid future CNI debugging.

### CODE REVIEW

File: client/allocrunner/networking_cni.go [hunk 1]
```diff
@@ -134,6 +140,22 @@ func (c *cniNetworkConfigurator) Setup(ctx context.Context, alloc *structs.Alloc
 			netStatus.Address = iface.IPConfigs[0].IP.String()
 		}
 	}
+
+	if netStatus.Address == "" {
+		c.logger.Debug("no address found for sandboxed interface from CNI result, using first available")
+		var found bool
+		for _, iface := range res.Interfaces {
+			if len(iface.IPConfigs) > 0 {
```
Comment (schmichael) [inline client/allocrunner/networking_cni.go:148]: Do we need the nil check from above as well?

File: client/allocrunner/networking_cni.go [hunk 2]
```diff
@@ -134,6 +140,22 @@ func (c *cniNetworkConfigurator) Setup(ctx context.Context, alloc *structs.Alloc
...
+				netStatus.Address = iface.IPConfigs[0].IP.String()
```
Comment (schmichael) [inline client/allocrunner/networking_cni.go:149]: If a sandbox interface is found above but it lacks an address, we will continue to use its name while we use another interfaces address.

If using one interfaces name and a different interfaces address is intentional, can we comment here?

**Review by schmichael — COMMENTED** [review]
Is it possible to unit test this? Maybe split the `cni.Setup` code and the address-finding-code into separate funcs as the latter seems unittestable. I'm actually less worried about correctness than I am about having a test as a sort of documentation for intended behavior. The "use first IP address we find" seems like strange behavior to me as someone who is not familiar with CNI plugin behavior.

**Comment by schmichael** [comment]
I refactored the CNIResult -> AllocNetworkStatus conversion into a new func for unit testing. I'd love to test any more configurations if you have them handy!

Sadly its still possible to get nondeterministic results from this new func, but I think the interface precedence logic is correct.

I think I may error in more cases than before.

Before we would only error if `CNIResults.Interfaces = map[string]*Interface{"some name": nil}`.

No we *also* error if there are no interfaces with an IP address set. I _think_ that was the intention of the original code, but not the actual behavior.

**Comment by nickethier** [comment]
Thanks for digging into this @schmichael I approve of these changes (even though I can't approve my own PR).

**Review by schmichael — APPROVED** [review]
Approving my own changes since @nickethier can't approve his own PR. :sweat_smile:

**Pull request merged**